platforms/common/uORB/uORBDeviceNode.cpp

```diff
@@ -73,210 +242,413 @@ static inline uint8_t round_pow_of_two_8(uint8_t n)
-		unlock();
+	orb_advert_t handle;
+	handle = nodeOpen(id, instance, true);
 
-		/* now complete the open */
-		return CDev::open(filp);
+	if (orb_advert_valid(handle)) {
+		node(handle)->advertise(publisher, queue_size);
 	}
 
-	/* is this a new subscriber? */
-	if (filp->f_oflags == PX4_F_RDONLY) {
+	return handle;
+}
 
-		/* allocate subscriber data */
-		SubscriptionInterval *sd = new SubscriptionInterval(_meta, 0, _instance);
+int uORB::DeviceNode::advertise(bool publisher, uint8_t queue_size)
+{
+	int ret = -1;
 
-		if (nullptr == sd) {
-			return -ENOMEM;
-		}
+	ret = ++_advertiser_count;
 
-		filp->f_priv = (void *)sd;
+	if (publisher) {
+		ret = ++_publisher_count;
+	}
 
-		int ret = CDev::open(filp);
+	update_queue_size(queue_size);
 
-		if (ret != PX4_OK) {
-			PX4_ERR("CDev::open failed");
-			delete sd;
-		}
+	return ret;
+}
 
-		return ret;
+int uORB::DeviceNode::orb_unadvertise(orb_advert_t &handle, bool publisher)
+{
+	int ret = -1;
+
+	if (orb_advert_valid(handle)) {
+		ret = node(handle)->unadvertise(publisher);
+		nodeClose(handle);
 	}
 
-	if (filp->f_oflags == 0) {
-		return CDev::open(filp);
+	return ret;
+}
+
+int uORB::DeviceNode::unadvertise(bool publisher)
+{
+	int ret = -1;
+
+	ret = --_advertiser_count;
+
+	if (publisher) {
+		--_publisher_count;
 	}
 
-	/* can only be pub or sub, not both */
-	return -EINVAL;
+	return ret;
 }
 
-int
-uORB::DeviceNode::close(cdev::file_t *filp)
+uORB::DeviceNode::DeviceNode(const ORB_ID id, const uint8_t instance, const char *path) :
+	_orb_id(id),
+	_instance(instance)
 {
-	if (filp->f_oflags == PX4_F_RDONLY) { /* subscriber */
-		SubscriptionInterval *sd = filp_to_subscription(filp);
-		delete sd;
+#if defined(CONFIG_BUILD_FLAT)
+	_devname = strdup(path);
+#else
+
+	if (strnlen(path, sizeof(_devname)) == sizeof(_devname)) {
+		PX4_ERR("node path too long %s", path);
+	}
+
+	strncpy(_devname, path, sizeof(_devname));
+#endif
+
+	int ret = px4_sem_init(&_lock, 1, 1);
+
+	if (ret != 0) {
+		PX4_DEBUG("SEM INIT FAIL: ret %d", ret);
+	}
+
+#ifndef CONFIG_BUILD_FLAT
+
+	_wait_item_freelist.set_item_pool(_wait_item_pool);
+
+	for (unsigned i = 0; i < sizeof(_wait_item_pool) / sizeof(_wait_item_pool[0]); i++) {
+		_wait_item_freelist.push(i);
 	}
 
-	return CDev::close(filp);
+#endif
 }
 
-ssize_t
-uORB::DeviceNode::read(cdev::file_t *filp, char *buffer, size_t buflen)
+uORB::DeviceNode::~DeviceNode()
 {
-	/* if the caller's buffer is the wrong size, that's an error */
-	if (buflen != _meta->o_size) {
-		return -EIO;
+	px4_sem_destroy(&_lock);
+
+#if defined(CONFIG_BUILD_FLAT)
+
+	// Delete all the allocated callback items.
+	// There should not be any left in the _callbacks queue since the node is
+	// deleted only if there are no more publishers or subscribers registered
+	while (!_wait_item_freelist.empty()) {
+		delete (static_cast<EventWaitItem *>(_wait_item_freelist.pop()));
 	}
 
-	return filp_to_subscription(filp)->copy(buffer) ? _meta->o_size : 0;
+	free(_devname);
```

Comment:
I can't find where this buffer is allocated, in the class header somewhere?

Reply:
it is from strdup; restricted to CONFIG_BUILD_FLAT. I don't even know how nuttx handles strdup freeing in PROTECTED nowdays (since by standard you must deallocate it with "free").